File: sport-ecommerce-fe/src/app/features/orders/services/order.service.ts
import { Injectable, inject } from '@angular/core';
import { Observable, map } from 'rxjs';

import { BaseHttpService } from '../../../core/services/base-http.service';
import { ORDER_API, ADMIN_ORDER_API } from '../../../core/constants/api-path.constant';
import {
  CheckoutRequest,
  OrderResponse,
  OrderSummaryResponse,
  OrderStatusUpdateRequest,
} from '../../../models/order.model';
import { ApiResponse } from '../../../models/api-response.model';
import { PageApiResponse } from '../../../models/page-response.model';

@Injectable({ providedIn: 'root' })
export class OrderService {
  private readonly http = inject(BaseHttpService);

  // ── Customer ──────────────────────────────────────────────────────────────

  placeOrder(request: CheckoutRequest): Observable<ApiResponse<OrderResponse>> {
    return this.http.post<OrderResponse>(ORDER_API.BASE, request);
  }

  getMyOrders(page = 0, size = 10): Observable<PageApiResponse<OrderSummaryResponse>> {
    return this.http.getPaged<OrderSummaryResponse>(
      ORDER_API.BASE,
      { page, size },
    );
  }

  getOrderById(id: number): Observable<ApiResponse<OrderResponse>> {
    return this.http.get<OrderResponse>(ORDER_API.BY_ID(id));
  }

  getOrderByNumber(orderNumber: string): Observable<ApiResponse<OrderResponse>> {
    return this.http.get<OrderResponse>(ORDER_API.BY_NUMBER(orderNumber));
  }

  cancelOrder(id: number): Observable<ApiResponse<OrderResponse>> {
    return this.http.delete<OrderResponse>(ORDER_API.CANCEL(id));
  }

  // ── Admin ─────────────────────────────────────────────────────────────────

  getAllOrders(page = 0, size = 20, status?: string): Observable<PageApiResponse<OrderSummaryResponse>> {
    return this.http.getPaged<OrderSummaryResponse>(
      ADMIN_ORDER_API.BASE,
      { page, size },
      status ? { status } : undefined,
    );
  }

  updateOrderStatus(id: number, request: OrderStatusUpdateRequest): Observable<ApiResponse<OrderResponse>> {
    return this.http.patch<OrderResponse>(ADMIN_ORDER_API.STATUS(id), request);
  }
}

File: sport-ecommerce-be/src/main/java/com/sport/ecommerce/modules/order/service/OrderService.java
package com.sport.ecommerce.modules.order.service;

import com.sport.ecommerce.modules.order.dto.request.CheckoutRequest;
import com.sport.ecommerce.modules.order.dto.request.OrderStatusUpdateRequest;
import com.sport.ecommerce.modules.order.dto.response.OrderResponse;
import com.sport.ecommerce.modules.order.dto.response.OrderSummaryResponse;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

public interface OrderService {

    /**
     * Creates an order from the current user's cart.
     * Deducts stock (pessimistic lock), clears the cart, persists the order.
     */
    OrderResponse placeOrder(CheckoutRequest request);

    /** Returns the authenticated user's orders, newest first. */
    Page<OrderSummaryResponse> getMyOrders(Pageable pageable);

    /** Returns a single order; throws 403 if the caller doesn't own it (unless admin). */
    OrderResponse getOrderById(Long id);

    /** Returns a single order by orderNumber (used on confirmation page). */
    OrderResponse getOrderByNumber(String orderNumber);

    /** Customer can cancel a PENDING order. */
    OrderResponse cancelOrder(Long id);

    // ── Admin ─────────────────────────────────────────────────────────────────

    Page<OrderSummaryResponse> getAllOrders(String status, Pageable pageable);

    OrderResponse updateStatus(Long id, OrderStatusUpdateRequest request);
}
